saleor/product/utils/attributes.py

```diff
@@ -1,70 +1,66 @@
+        attributes_display.append(", ".join(translated_values))
 
-def get_attributes_display_map(obj, attributes):
-    """Return attributes associated with an object, as dict of AttrPK: AttributeValue.
+    return " / ".join(attributes_display)
 
-    Args:
-        obj: The variant or product.
-        attributes: Attribute Iterable
 
-    """
-    display_map = defaultdict(str)
-    for attribute in attributes:
-        attribute_values = obj.attributes.get(str(attribute.pk))  # type: List
-        if attribute_values:
-            choices = {str(a.pk): a.translated for a in attribute.values.all()}
-            for value in attribute_values:
-                current_display_value = display_map[attribute.pk]
-                if not current_display_value:
-                    current_display_value = choices[value]
-                else:
-                    current_display_value = f"{current_display_value}, {choices[value]}"
-                display_map[attribute.pk] = current_display_value
-    return display_map
+def _associate_attribute_to_instance(
+    instance: Union[Product, ProductVariant], attribute_pk: Attribute
+) -> T_ASSIGNMENT_REL:
+    """Associate a given attribute to an instance."""
+    if isinstance(instance, Product):
+        attribute_rel = instance.product_type.attributeproduct.get(
+            attribute_id=attribute_pk
+        )
+
+        assignment, _ = AssignedProductAttribute.objects.get_or_create(
+            product=instance, assignment=attribute_rel
+        )
+    elif isinstance(instance, ProductVariant):
+        attribute_rel = instance.product.product_type.attributevariant.get(
+            attribute_id=attribute_pk
+        )
 
+        assignment, _ = AssignedVariantAttribute.objects.get_or_create(
+            variant=instance, assignment=attribute_rel
+        )
+    else:
+        raise AssertionError(f"{instance.__class__.__name__} is unsupported")
 
-def generate_name_from_values(attributes_dict):
-    """Generate name from AttributeValues.
+    return assignment
 
-    Attributes dict is sorted, as attributes order should be kept within each save.
 
-    Args:
-        attributes_dict: dict of attribute_pk: AttributeValue values
+def associate_attribute_values_to_instance(
+    instance: Union[Product, ProductVariant],
+    attribute: Attribute,
+    *values: AttributeValue,
+) -> T_ASSIGNMENT_REL:
+    """Assign given attribute values to a product or variant.
 
+    Note: be award this function invokes the ``set`` method on the instance's
+    attribute association. Meaning any values already assigned or concurrently
+    assigned will be overridden by this call.
     """
-    return " / ".join(
-        str(attribute_value)
-        for attribute_pk, attribute_value in sorted(
-            attributes_dict.items(), key=lambda x: x[0]
-        )
-    )
+    assignment = _associate_attribute_to_instance(instance, attribute.pk)
+    assignment.values.set(values)
```

Comment:
Does it allow to assign values of a different attribute than the one that was provided or does the relationship protects us from it?

Reply:
Yes, if one calls this directly with wrong values, it will still assign them. But it is the job of `AttributeAssignmentMixin` to check whether the value is from this attribute or not, otherwise it would really be heavy to check here.

Reply:
I guess we could just iterate over `values` and see if `value.attribute` is the same as the one passed as an argument. That could work if the `values` are saved to the database at this point, not sure if they are.

Reply:
I'm still thinking what's the best way to keep the validation rules in Saleor, which are part of the business logic. I think that in general API layer should just call the right methods and validation should happen in some lower layer, like this. It's something to discuss.